src/DataProtection/DataProtection/src/KeyManagement/XmlKeyManager.cs

```diff
@@ -286,6 +299,18 @@ public CancellationToken GetCacheExpirationToken()
         }
     }
 
+    private void RecordKey(Key key)
+    {
+        if (!_knownKeyMap.TryAdd(key.KeyId, key))
+        {
+            // If we lost the race, the winner inserted an equivalent key
+            Debug.Assert(_knownKeyMap.TryGetValue(key.KeyId, out var existingKey));
+            Debug.Assert(existingKey.CreationDate == key.CreationDate);
+            Debug.Assert(existingKey.ActivationDate == key.ActivationDate);
+            Debug.Assert(existingKey.ExpirationDate == key.ExpirationDate);
```

Comment:
Do we need to check the revocation status here? Also, I assume that in all cases none of these assertions would fail?

Reply:
I'm pretty sure I consciously skipped `IsRevoked` because it's mutable, but it's been too long.

Reply:
In particular, I don't think that gets set until after RecordKey is called.  Normally, caching mutable values is bad, but it can only become more revoked, so I think it's fine.

Reply:
Even better, it looks like we clone values retrieved from the cache, so they _are_ immutable.  And it doesn't matter if they're subsequently revoked because that mutation happens after the cache retrieval and cloning.

Reply:
To, TL;DR: you can't check revocation status, since it could change and no, these assertions should never fail because you can't mutate those properties (and key IDs shouldn't collide).